packages/edit-site/src/store/actions.js

```diff
@@ -257,22 +257,48 @@ export function setEditedPostContext( context ) {
 export const setPage =
 	( page ) =>
 	async ( { dispatch, registry } ) => {
-		if ( ! page.path && page.context?.postId ) {
-			const entity = await registry
+		let template;
+		const getDefaultTemplate = async ( slug ) =>
+			apiFetch( {
+				path: addQueryArgs( '/wp/v2/templates/lookup', {
+					slug: `page-${ slug }`,
+				} ),
+			} );
+
+		if ( page.path ) {
+			template = await registry
+				.resolveSelect( coreStore )
+				.__experimentalGetTemplateForLink( page.path );
+		} else {
+			const editedEntity = await registry
 				.resolveSelect( coreStore )
-				.getEntityRecord(
+				.getEditedEntityRecord(
 					'postType',
-					page.context.postType || 'post',
-					page.context.postId
+					page.context?.postType || 'post',
+					page.context?.postId
 				);
-			// If the entity is undefined for some reason, path will resolve to "/"
-			page.path = getPathAndQueryString( entity?.link );
+			const currentTemplateSlug = editedEntity?.template;
+			if ( currentTemplateSlug ) {
+				const currentTemplate = (
+					await registry
+						.resolveSelect( coreStore )
+						.getEntityRecords( 'postType', 'wp_template', {
+							per_page: -1,
+						} )
+				 )?.find( ( { slug } ) => slug === currentTemplateSlug );
+				if ( currentTemplate ) {
+					template = currentTemplate;
+				} else {
+					// If a page has a `template` set and is not included in the list
+					// of the current theme's templates, query for current theme's default template.
+					template = await getDefaultTemplate( editedEntity?.link );
+				}
+			} else {
+				// Page's `template` is empty, that indicates we need to use the default template for the page.
+				template = await getDefaultTemplate( editedEntity?.link );
+			}
 		}
 
-		const template = await registry
-			.resolveSelect( coreStore )
-			.__experimentalGetTemplateForLink( page.path );
-
 		if ( ! template ) {
 			return;
 		}
```

Comment:
I think the change in this action is fine, but what would be awesome (and probably better explored as a follow-up) is to remove the need for this action in favor of a `selector + resolver` that contains the same logic.

In other words when the editor component would calls these selectors:

```
getEditedEntityId()
getEditedEntityType()
```

a resolver is triggered to actually "resolve" the template (so the logic we have right now in this action).

The reason for this is that we would be able to say that `editEntityRecord` of the `template` actually marks the resolver as "unresolved" which would cause the resolver to trigger again. 

There are some subtle things to consider (loading...) that's why this is more of a quality of life improvement that should be addressed separately.

Reply:
>The reason for this is that we would be able to say that editEntityRecord of the template actually marks the resolver as "unresolved" which would cause the resolver to trigger again.

Can you expand on this a bit?

Reply:
I mean that we can define the `shouldInvalidate` property on the resolver to ensure that every time the edit entity action is called with a new template, the corresponding resolver is invalidated.